### PR TITLE
Add disable argument in DropdownMenuItem

### LINKED ISSUE #75609: Allow DropDownMenuItem to be disabled.
Have a property for DropDownMenuItem to be disabled or not

Doing this, item wont be selected if user selects on a disabled item.

This is very useful in showing tree like dropdown with parent and child where parent cannot be clicked.

### PR DESCRIPTION
**Pull request by werainkhatri**
Adds a disable argument in `DropdownMenuItem` widget and added tests.
Design Doc: http://flutter.dev/go/disable-dropdownmenuitem

Fixes: #75609 #75487 

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*
I ran the tests locally and it showed that there is 1 failure, but I could not find which one.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

One of the existing tests were not passing. This was due to change in wording of an assertion, and the test checked for the assertion words. I fixed it by changing the checked phrase to the corrected one.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/material/dropdown.dart [hunk 1]
```diff
@@ -686,17 +689,26 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
     Key? key,
     this.onTap,
     this.value,
+    this.disabled = false,
     required Widget child,
-  }) : assert(child != null),
-       super(key: key, child: child);
+  })   : assert(child != null),
+        super(key: key, child: child);
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/dropdown.dart:695]: Revert unrelated lines of code

File: packages/flutter/lib/src/material/dropdown.dart [hunk 2]
```diff
@@ -686,17 +689,26 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
...
   ///
   /// Eventually returned in a call to [DropdownButton.onChanged].
   final T? value;
+
+  /// If `true`:
+  /// * [value] and [onTap] won't be used.
+  /// * the item will not be clickable / focusable.
+  /// * tapping won't close drop down list.
+  ///
+  /// defaults to `false`
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/dropdown.dart:710]: We usually have the following structure for API docs:

/// A single line to succinctly describe the parameter or API.
///
/// More content to further elaborate on the API itself. This can span multiple
/// lines to explain the purpose of the API and how to use it in a more elaborate
/// way.
///
/// Sometimes, describing what the default behavior is. (ie. If `null`, then [parameter] is used).
///
/// Sometimes, a `See more:` section (see other API docs for an idea of how this looks like)

File: packages/flutter/test/material/dropdown_test.dart [hunk 3]
```diff
@@ -3052,17 +3052,180 @@ void main() {
...
+          matching: find.byType(CustomPaint),
+        ).at(2),
+        paints
+          ..save()
+          ..rrect()
+          ..rrect()
+          ..rrect()
         // The height of menu is 47.0.
-        ..rrect(rrect: const RRect.fromLTRBXY(0.0, 0.0, 112.0, 47.0, 2.0, 2.0), color: Colors.grey[50], hasMaskFilter: false)
+          ..rrect(rrect: const RRect.fromLTRBXY(0.0, 0.0, 112.0, 47.0, 2.0, 2.0), color: Colors.grey[50], hasMaskFilter: false)
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3065]: Revert unrelated LOC

File: packages/flutter/test/material/dropdown_test.dart [hunk 4]
```diff
@@ -3052,17 +3052,180 @@ void main() {
...
+
+    // Tap on a disabled item.
+    await tester.tap(find.text('disabled').hitTestable());
+    await tester.pumpAndSettle();
+
+    // The dropdown should still be open, i.e., there should be one widget with 'second' text.
+    expect(find.text('second').hitTestable(), findsOneWidget);
+  });
+
+  testWidgets('disabled item should not be focusable when tapped', (WidgetTester tester) async {
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3103]: I think the correct test here would show that hovering over the disabled item (in web and desktop) keeps focus on original item.

The reason being that tapping on an (enabled) item does not focus the item, but instead closes the dropdown and selects it in the dropdown menu.

File: packages/flutter/test/material/dropdown_test.dart [hunk 5]
```diff
@@ -3052,17 +3052,180 @@ void main() {
...
+    expect(Focus.of(tester.element(find.text('second').hitTestable())).hasPrimaryFocus, isTrue);
+
+    // Moving focus one up.
+    await tester.sendKeyEvent(LogicalKeyboardKey.arrowUp);
+    await tester.pump();
+    // now first should have the focus.
+    expect(Focus.of(tester.element(find.text('first').hitTestable())).hasPrimaryFocus, isTrue);
+  });
+
+  testWidgets('an enabled and a disabled item can have the same value', (WidgetTester tester) async {
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3193]: Maybe change this to check to see if disabled items are correctly ignored by the menu system (for example, if there are 10 disabled items with "value" defined). 

On the other hand, maybe we do not want to allow disabled items to have values in the first place, since it does not provide it with any useful meaning. Also, this can easily be a place for developer to write bugs in their code (ie. they dynamically create their menu items, and somehow they write incorrect logic for value generation and deciding if the item is disabled/enabled. This would give them quite the headache to debug).

File: packages/flutter/test/material/dropdown_test.dart (hunk 4 above)
Comment (werainkhatri) [reply packages/flutter/test/material/dropdown_test.dart:3103]: I understand and agree with your point. Can you tell me how to do that? How do I simulate focusing / hovering over the disabled item in test.

File: packages/flutter/test/material/dropdown_test.dart (hunk 5 above)
Comment (werainkhatri) [reply packages/flutter/test/material/dropdown_test.dart:3193]: I have added assertions so that [value] and [onTap] need to be `null` if [disabled] is `true`

File: packages/flutter/lib/src/material/dropdown.dart (hunk 2 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart:710]: Is this okay?
```dart
/// Whether a user can select this menu item. Defaults to `false`.
///
/// If `true`: [value] and [onTap] should be `null`
/// and user will **not** be able to select this item.
///
/// if `false`: user will be able to select this item.
```

Now I feel the name of the variable should be [enabled] or [isEnabled] instead of [disabled]. What do you say?

Comment (shihaohong) [reply packages/flutter/lib/src/material/dropdown.dart]: `enabled` or `isEnabled` is fine too.  Remember, the first sentence should only be one line. Here are some suggestions.
```
/// Whether or not a user can select this menu item. 
/// 
/// If `true`, [value] and [onTap] should be `null` and the user 
/// will **not** be able to select this item. If `false`, then the user 
/// will be able to select this item.
/// 
/// Defaults to `false`.
```

Of course, change the wording based on whether you flip the property name to `enabled` or not.

Comment (shihaohong) [reply packages/flutter/lib/src/material/dropdown.dart]: You may also want to consider explaining why `value` and `onTap` should be `null` here.

File: packages/flutter/test/material/dropdown_test.dart (hunk 4 above)
Comment (shihaohong) [reply packages/flutter/test/material/dropdown_test.dart]: Here's a pointer to some hover tests. There are a bunch of others in the repo you can reference: https://github.com/flutter/flutter/blob/f835839f6b752a62a6f0463e51a64f20c4463982/packages/flutter/test/material/bottom_navigation_bar_test.dart#L1811

File: packages/flutter/lib/src/material/dropdown.dart [hunk 6]
```diff
@@ -692,17 +695,30 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
...
     required Widget child,
-  }) : assert(child != null),
-       super(key: key, child: child);
+  })   : assert(child != null),
+          assert(!disabled || (disabled && value == null), 'if disabled is true, value should be null'),
+          assert(!disabled || (disabled && onTap == null), 'if disabled is true, onTap should be null'),
+        super(key: key, child: child);
 
   /// Called when the dropdown menu item is tapped.
+  /// If disabled is `true`, it should be `null`.
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/dropdown.dart:706]:
```suggestion
  ///
  /// If disabled is `true`, it should be `null`.
```

File: packages/flutter/lib/src/material/dropdown.dart [hunk 7]
```diff
@@ -692,17 +695,30 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
...
   final VoidCallback? onTap;
 
   /// The value to return if the user selects this menu item.
+  /// If disabled is `true`, it should be `null`.
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/dropdown.dart:710]:
```suggestion
  ///
  /// If disabled is `true`, it should be `null`.
```

File: packages/flutter/test/material/dropdown_test.dart (hunk 4 above)
Comment (werainkhatri) [reply packages/flutter/test/material/dropdown_test.dart]: I have added a test in the recently reviewed commit, to see if a disabled item is even focusable at all.
So should I add this hover test? Because if the disabled item is not within a focus (as verified by that test) it won't get the focus when hovered either.

File: packages/flutter/lib/src/material/dropdown.dart (hunk 2 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart]: Is this okay?
```dart
/// Whether or not a user can select this menu item.
///
/// If `false`, the user will be unable to select this item,
/// and since [value] and [onTap] are redundant for a disabled
/// and unselectable menu item, both should be `null`.
/// If `true`, the user will be able to select this item.
///
/// Defaults to `true`
```

File: packages/flutter/lib/src/material/dropdown.dart (hunk 1 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart]: Done

File: packages/flutter/lib/src/material/dropdown.dart (hunk 6 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart]: Done

File: packages/flutter/lib/src/material/dropdown.dart (hunk 7 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart]: Done

File: packages/flutter/test/material/dropdown_test.dart [hunk 8]
```diff
@@ -3122,4 +3122,156 @@ void main() {
...
+                DropdownMenuItem<String>(
+                  value: 'first',
+                  child: Text('first'),
+                ),
+                DropdownMenuItem<String>(
+                  value: 'second',
+                  child: Text('second'),
+                ),
+              ],
+              onChanged: (String? v) => setState(() => value = v),
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3149]: 1) Avoid using single letter variable names. Prefer to describe what the variables are
2) The fat arrow function (=>) implies that the value is being returned. This could be confusing at first glance since the return value does not matter here (void callback)
```suggestion
              onChanged: (String? newValue) {
                setState(() {
                  value = newValue;
                });
              },
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 9]
```diff
@@ -3122,4 +3122,156 @@ void main() {
...
+      );
+    }
+  });
+
+  testWidgets('value should be null if enabled is false', (WidgetTester tester) async {
+    final List<String> values = <String>['enabled', 'disabled'];
+
+    try {
+      // trying to assign [value] and [onTap] non-null values irrespective of
+      // if the menu item is disabled or now
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3213]:
```suggestion
      // Try to assign [value] and [onTap] non-null values irrespective of
      // if the menu item is disabled for now
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 10]
```diff
@@ -3122,4 +3122,156 @@ void main() {
...
+      await tester.pumpWidget(
+        MaterialApp(
+          home: Scaffold(
+            body: DropdownButton<String>(
+              value: values[0],
+              items: values.map<DropdownMenuItem<String>>((String e) => DropdownMenuItem<String>(
+                value: e,
+                child: Text(e),
+                enabled: e == 'enabled',
+              )).toList(),
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3223]: Maybe:
```suggestion
              items: values.map<DropdownMenuItem<String>>((String value) => DropdownMenuItem<String>(
                value: value,
                child: Text(value),
                enabled: value == 'enabled',
              )).toList(),
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 11]
```diff
@@ -3122,4 +3122,156 @@ void main() {
...
+      );
+    }
+  });
+
+  testWidgets('onTap should be null if enabled is false', (WidgetTester tester) async {
+    final List<String> values = <String>['enabled', 'disabled'];
+
+    try {
+      // trying to assign [value] and [onTap] non-null values irrespective of
+      // if the menu item is disabled or now
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3245]:
```suggestion
      // Try to assign [value] and [onTap] non-null values irrespective of
      // if the menu item is disabled for now
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 12]
```diff
@@ -3122,4 +3122,156 @@ void main() {
...
+      await tester.pumpWidget(
+        MaterialApp(
+          home: Scaffold(
+            body: DropdownButton<String>(
+              value: values[0],
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3250]:
```suggestion
              value: 'enabled',
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 13]
```diff
@@ -3122,4 +3122,156 @@ void main() {
...
+    } on AssertionError catch (error) {
+      expect(
+        error.toString(),
+        contains('if enabled is false, value should be null'),
+      );
+    }
+  });
+
+  testWidgets('onTap should be null if enabled is false', (WidgetTester tester) async {
+    final List<String> values = <String>['enabled', 'disabled'];
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3241]: This isn't really used:
```suggestion
```

**Comment by werainkhatri** [comment]
@shihaohong Any updates on this PR?

File: packages/flutter/test/material/dropdown_test.dart [hunk 14]
```diff
@@ -3122,4 +3122,162 @@ void main() {
...
+                ),
+                DropdownMenuItem<String>(
+                  value: 'second',
+                  child: Text('second'),
+                ),
+              ],
+              onChanged: (String? newValue) {
+                setState(() {
+                value = newValue;
+              });
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3152]: nit indentation
```suggestion
              onChanged: (String? newValue) {
                setState(() {
                  value = newValue;
                });
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 15]
```diff
@@ -3122,4 +3122,162 @@ void main() {
...
+                  enabled: false,
+                  child: const Text('disabled'),
+                  onTap: () {},
+                ),
+                const DropdownMenuItem<String>(
+                  value: 'enabled',
+                  child: Text('enabled'),
+                )
+              ],
+              onChanged: onChanged,
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3268]:
```suggestion
            body: DropdownButton<String>(
              value: 'enabled',
              onChanged: onChanged,
              items: <DropdownMenuItem<String>>[
                DropdownMenuItem<String>(
                  enabled: false,
                  child: const Text('disabled'),
                  onTap: () {},
                ),
                const DropdownMenuItem<String>(
                  value: 'enabled',
                  child: Text('enabled'),
                )
              ],
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 16]
```diff
@@ -3122,4 +3122,162 @@ void main() {
...
+                  value: 'disabled',
+                  enabled: false,
+                  child: const Text('disabled'),
+                ),
+                const DropdownMenuItem<String>(
+                  value: 'enabled',
+                  child: Text('enabled'),
+                )
+              ],
+              onChanged: onChanged,
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3232]: nit: prefer to avoid trailing params since they might be missed
```suggestion
              value: 'enabled',
              onChanged: onChanged,
              items: <DropdownMenuItem<String>>[
                DropdownMenuItem<String>(
                  value: 'disabled',
                  enabled: false,
                  child: const Text('disabled'),
                ),
                const DropdownMenuItem<String>(
                  value: 'enabled',
                  child: Text('enabled'),
                )
              ],
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 17]
```diff
@@ -3122,4 +3122,162 @@ void main() {
...
+              DropdownMenuItem<String>(
+                enabled: false,
+                child: Text('disabled'),
+              ),
+              DropdownMenuItem<String>(
+                value: 'enabled',
+                child: Text('enabled'),
+              )
+            ],
+            onChanged: onChanged,
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3188]:
```suggestion
          body: DropdownButton<String>(
            value: 'enabled',
            onChanged: onChanged,
            items: const <DropdownMenuItem<String>>[
              DropdownMenuItem<String>(
                enabled: false,
                child: Text('disabled'),
              ),
              DropdownMenuItem<String>(
                value: 'enabled',
                child: Text('enabled'),
              )
            ],
```

File: packages/flutter/lib/src/material/dropdown.dart [hunk 18]
```diff
@@ -692,17 +695,34 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
     Key? key,
     this.onTap,
     this.value,
+    this.enabled = true,
     required Widget child,
   }) : assert(child != null),
-       super(key: key, child: child);
+        assert(enabled || (!enabled && value == null), 'if enabled is false, value should be null'),
+        assert(enabled || (!enabled && onTap == null), 'if enabled is false, onTap should be null'),
+        super(key: key, child: child);
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/dropdown.dart:703]:
```suggestion
       assert(enabled || (!enabled && value == null), 'if enabled is false, value should be null'),
       assert(enabled || (!enabled && onTap == null), 'if enabled is false, onTap should be null'),
       super(key: key, child: child);
```

**Review by shihaohong — APPROVED** [review]
Tiny nits here and there about styling. I'll commit them for you, but take note for future PRs. 

Thank you so much for the contribution!

File: packages/flutter/lib/src/material/dropdown.dart (hunk 18 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart]: Is there a tool or some method to indent lines correctly or am I supposed to keep in mind the indentation while coding?

**Comment by werainkhatri** [comment]
> Tiny nits here and there about styling. I'll commit them for you, but take note for future PRs.
> 
> Thank you so much for the contribution!

I'll keep these in mind.

Thank you so so much for your time guiding me into writing better more efficient code. :)

**Comment by werainkhatri** [comment]
The `flutter-gold` check is pending since 4 days. Ig there is some problem. Is there something I can do?

**Comment by shihaohong** [comment]
It seems like the Linux web tests are failing. Could you check to see what might be causing the failure? It could also be a flake.

**Comment by werainkhatri** [comment]
> It seems like the Linux web tests are failing. Could you check to see what might be causing the failure? It could also be a flake.

One of the tests, [link](https://www.github.com/flutter/flutter/tree/3db5d413c9436704d0e174198c25d4be59cda335/packages%2Fflutter%2Ftest%2Fmaterial%2Fdropdown_test.dart#L3215), is failing.
In this test, I am checking if a widget is focusable by checking if it has a FocusScope as a parent (by trying to get `Focus.of(*widget*)` which throws an error if the widget has no focus scope in it's parent tree) and catching the thrown error.

**Comment by werainkhatri** [comment]
@shihaohong I have fixed the bug. 
Could you review it and add the `waiting for tree to go green` label?

File: packages/flutter/lib/src/material/dropdown.dart [hunk 19]
```diff
@@ -681,21 +684,40 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
   /// Creates an item for a dropdown menu.
   ///
   /// The [child] argument is required.
+  ///
+  /// If [enabled] is `false`, [value] and [onTap] should be null
```
Comment (shihaohong) [inline packages/flutter/lib/src/material/dropdown.dart:688]: nit: 
```suggestion
  /// If [enabled] is `false`, [value] and [onTap] should be null.
```

File: packages/flutter/test/material/dropdown_test.dart [hunk 20]
```diff
@@ -3165,4 +3165,164 @@ void main() {
...
+    await tester.tap(find.text('enabled').hitTestable());
+    await tester.pumpAndSettle();
+
+    try {
+      // try to get the focus node for disabled item
+      final FocusNode node = Focus.of(tester.element(find.text('disabled').hitTestable()));
+      // For disabled item, [Focus.of] should throw an exception,
+      // but it doesn't when platform=chrome and web-renderer=html when flutter test is run
+      // In that case, the node should not be able to request focus
+      expect(node.canRequestFocus, false, reason: 'Disabled menu item should not be able to request focus');
```
Comment (shihaohong) [inline packages/flutter/test/material/dropdown_test.dart:3247]: Interesting -- so it does not throw an error only for web, but the `FocusNode` itself is set to not being able to request focus? Do you happen to know the reason this is the case?

cc/ @gspencergoog for comment, since he's very familiar with focus-related stuff!

Comment (werainkhatri) [reply packages/flutter/test/material/dropdown_test.dart:3247]: Yes, precisely.
I have tried to understand the reason to the problem, but it was difficult to understand all the underlying modules, being a beginner. I have asked out for it in the flutter discord and searched over in the flutter docs but have not got anything yet.

Comment (gspencergoog) [reply packages/flutter/test/material/dropdown_test.dart:3247]: `Focus.of` retrieves the focus node for the nearest `Focus` widget, and `FocusScope.of` retrieves the focus node for the nearest `FocusScope` widget. The `Focus.of` can also request either one by supplying `scopeOk: true`.

Also, you can test for existence of such a node without having to catch the assert, by just using `Focus.maybeOf` or `FocusScope.maybeOf` instead, and checking for null.

So, the test for this should look like:

```dart
Element disabled = tester.element(find.text('disabled').hitTestable()));
expect(Focus.maybeOf(disabled, scopeOk: true).canRequestFocus, false, reason: 'Disabled menu item should not be able to request focus');
```

I'm not sure why the web implementation might be failing, but you can dump the focus tree there with `debugDumpFocusTree();` to see what the hierarchy looks like. If that still doesn't give you enough information, you can turn on focus debugging by setting `_kDebugFocus = true;` at the top of [`focus_manager.dart`](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/widgets/focus_manager.dart;l=20;drc=f82046b150ad9d54962ad8098df2babdad34c596).

Comment (werainkhatri) [reply packages/flutter/test/material/dropdown_test.dart:3247]: I understood the problem from the focus tree. There is a focus node in the disabled item for web because of web shortcuts. Should I remove the web shortcuts for disabled item?

```dart
if (kIsWeb) {
  child = Shortcuts(
  shortcuts: _webShortcuts,
  child: child,
  );
}
```

Also, the web shortcuts function says the following ([link](https://github.com/flutter/flutter/blob/f0934fdee35613e4eaf576f460d2a3ff75a1962f/packages/flutter/lib/src/material/dropdown.dart#L150))
```dart
// On the web, up/down don't change focus, *except* in a <select>
// element, which is what a dropdown emulates.
```

<details>
<summary>web implementation focus tree</summary>

```dart
Child 2: FocusScopeNode#66697(_ModalScopeState<dynamic> Focus Scope [IN FOCUS PATH])
│ context: FocusScope
│ IN FOCUS PATH
│ focusedChildren: FocusNode#38cd3([PRIMARY FOCUS])
│
├─Child 1: FocusNode#ba582(Shortcuts) // disabled item
│   context: Focus
│   NOT FOCUSABLE
│
└─Child 2: FocusNode#3ec9f(Shortcuts [IN FOCUS PATH])
    │ context: Focus
    │ NOT FOCUSABLE
    │ IN FOCUS PATH
    │
    └─Child 1: FocusNode#38cd3([PRIMARY FOCUS]) // enabled item
           context: Focus
           PRIMARY FOCUS
```
</details>
<details>
<summary>non web implementation focus tree</summary>

```dart
Child 2: FocusScopeNode#5c8f2(_ModalScopeState<dynamic> Focus Scope [IN FOCUS PATH])
│ context: FocusScope
│ IN FOCUS PATH
│ focusedChildren: FocusNode#7972a([PRIMARY FOCUS])
│
└─Child 1: FocusNode#7972a([PRIMARY FOCUS]) // enabled item
        context: Focus
        PRIMARY FOCUS
```
</details>

Comment (gspencergoog) [reply packages/flutter/test/material/dropdown_test.dart:3247]: Yes, it seems reasonable to not include the shortcuts for a disabled item, as long as you can determine that it really is disabled at the point where the shortcuts are inserted.

Comment (werainkhatri) [reply packages/flutter/test/material/dropdown_test.dart]: I have added a constructor argument to ask from the user if it is enabled or not, if it's what you are asking for. If `enabled` is `false`, it is disabled.

**Review by shihaohong — APPROVED** [review]
LGTM. Digging into the web focus stuff was pretty interesting

File: packages/flutter/lib/src/material/dropdown.dart [hunk 21]
```diff
@@ -681,21 +684,40 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
...
+    this.enabled = true,
     required Widget child,
   }) : assert(child != null),
+       assert(enabled || (!enabled && value == null), 'if enabled is false, value should be null'),
+       assert(enabled || (!enabled && onTap == null), 'if enabled is false, onTap should be null'),
        super(key: key, child: child);
 
   /// Called when the dropdown menu item is tapped.
+  ///
+  /// If [enabled] is `false`, this property should be `null`.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/dropdown.dart:702]: Actually, in keeping with our other widgets, this should probably be what determines if it is enabled or not.  We typically disable items (buttons, etc.) when their onTap is null. That way you don't need a separate "enabled" accessor: it is sufficient to just set onTap to null.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 22]
```diff
@@ -681,21 +684,40 @@ class DropdownMenuItem<T> extends _DropdownMenuItemContainer {
...
   final VoidCallback? onTap;
 
   /// The value to return if the user selects this menu item.
   ///
   /// Eventually returned in a call to [DropdownButton.onChanged].
+  ///
+  /// If [enabled] is `false`, this property should be `null`.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/dropdown.dart:709]: Why is this a requirement? As long as it is disabled, the value is irrelevant.

**Review by gspencergoog — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/material/dropdown.dart (hunk 21 above)
Comment (shihaohong) [reply packages/flutter/lib/src/material/dropdown.dart:702]: One problem is that this would be a breaking change -- Currently, the way the `DropdownMenuItem` works is that defining `onTap` is a callback that is called on top of the behavior of selecting the item. This means that any user that hasn't defined their `onTap` callback will suddenly have their menu items be disabled.

File: packages/flutter/lib/src/material/dropdown.dart (hunk 22 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart:709]: This was suggested by @shihaohong sir in [this](https://github.com/flutter/flutter/pull/76968#discussion_r589275180) conversation. If you feel otherwise, I'll change it.

Comment (shihaohong) [reply packages/flutter/lib/src/material/dropdown.dart:709]: The rationale I had here was that it might be confusing to developers. For example, if someone had (unknowingly) disabled a menu item and then defined the `value` or `onTap`, then the irrelevant values are still defined in their code but have no effect on the widget itself, since it's disabled. I was thinking it might be better to catch this in an assert ahead of time than allow devs to scratch their heads trying to figure it out.

File: packages/flutter/lib/src/material/dropdown.dart (hunk 21 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart:702]: Considering the present implementation (on master branch) of `DropdownMenuItem`, `onTap` can be `null` and the item will still be enabled. Altering this code design (according to your suggestion) will be a breaking change.

Also, less important, but there may be a situation where the user doesn't want an `onTap` call on an enabled item. In this case the user is compelled to add a `VoidCallback` to ensure the item is enabled.

File: packages/flutter/lib/src/material/dropdown.dart (hunk 22 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/dropdown.dart:709]: Well, but it also means that they have to have extra logic. Instead of just setting a static value on the widget, they now need to have `enabled ? value : null` instead. Same with the onTap, although there it makes a little more sense to assert, since they might be confused as to why their onTap is never called when its disabled. It's still extra logic that they have to provide now, since it'll assert otherwise. They might not care that there's an onTap defined that is only called when enabled is true.

So, instead of what would be consistent with other widgets:
```dart
DropdownMenuItem(
  // ...
  onTap: myEnabled ? () { /*...*/ } : null,
  value: foo,
)
```

They have to write the (fairly redundant):
```dart
DropdownMenuItem(
  // ...
  onTap: myEnabled ? () { /*...*/ } : null,
  value: myEnabled ? foo : null,
  enabled: myEnabled,
)
```

And because of the asserts, they don't have the option of the more readable:
```
DropdownMenuItem(
  // ...
  onTap: () { /*...*/ },
  value: foo,
  enabled: myEnabled,
)
```

However, I realize that since we didn't have the consistent API from the beginning, it's pretty common to have a dropdown item with no `onTap` set now, so I guess we can't get away with the breaking change, but maybe we can skip the asserts?

Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart:709]: Instead of assigning the disability of the item with `onTap` being `null`, shouldn't it be done with `value` being `null`? i.e., if `value` is `null`, item will be disabled, else enabled. Right now, item's value being `null` has no significance. This would give it that, and would keep the pattern as with other widgets.

For `onTap`, since it is not directly related to the basic working of `DropdownButton`, we can excempt it from being dependent on item being enabled or not, i.e., `onTap` will work irrespective of item being disabled or enabled (or `value` being `null` or not).

Comment (shihaohong) [reply packages/flutter/lib/src/material/dropdown.dart:709]: > Instead of assigning the disability of the item with onTap being null, shouldn't it be done with value being null?

@werainkhatri That would actually be a breaking change because I recently worked through some issues with users expecting `null` values to be valid (so to these subset of users, `null` value do have significance).

Let's remove the asserts. I think @gspencergoog's concerns about users requiring additional logic in their code just to disable their dropdown menu items makes a lot of sense

Comment (werainkhatri) [reply packages/flutter/lib/src/material/dropdown.dart]: Removed the asserts.

I also came across a bug which I fixed.
I went through some other related PRs and realized that if `DropdownButton.value` is `null` and there is an item with `value` as null, that item is selected. But if there are >= 1 disabled items along with that, it causes an assertion error as they are considered similar. I changed the condition to selecting the item if its `value` is `null` and its not disabled.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by werainkhatri** [comment]
Could you help me correct the failing check (**Windows framwork_tests_misc**)? I don't think its related to my PR.

**Comment by gspencergoog** [comment]
I restarted it, and that didn't fix it, so you might need to `git pull --rebase`, and push again. If that still doesn't fix it, then we just have to wait until it's fixed on master (which should be pretty quicK).

**Comment by werainkhatri** [comment]
I have merged the latest master branch recently and there haven't been any changes since.
Also, I checked the latest merged PR #80474 in which all checks have passed.

**Comment by werainkhatri** [comment]
**Google testing** has failed
@shihaohong @gspencergoog

**Comment by gspencergoog** [comment]
It looks like an unrelated failure. Checking.
